DiagnosticLogs: Add CHIP_ERROR Parameter to Reset and EndLogCollection Methods

pimpalemahesh opened the pull request:
Fixes #36776 

### Change Overview:
1. Modified the Reset() method of BDXDiagnosticLogsProvider to include a CHIP_ERROR parameter.
2. Updated the EndLogCollection() method of DiagnosticLogsProviderDelegate to accept the same parameter.
3. Ensured that EndLogCollection() is called with CHIP_NO_ERROR for a successful BDX transfer and with the appropriate error code for failures.

pimpalemahesh committed:
BDX: update reset and endLogCollection to distinguish success and failure cases

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=36775) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=36775) before we can accept your contribution.<br/><hr/>**esp** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=36775) it.</sub>

pimpalemahesh commented:
@vivien-apple Please review

bzbarsky-apple commented on src/app/clusters/diagnostic-logs-server/DiagnosticLogsProviderDelegate.h:
```diff
@@ -59,9 +59,10 @@ class DiagnosticLogsProviderDelegate
      *
      */
-    virtual CHIP_ERROR EndLogCollection(LogSessionHandle sessionHandle) = 0;
+    virtual CHIP_ERROR EndLogCollection(LogSessionHandle sessionHandle, CHIP_ERROR error) = 0;
```
This is a backwards-incompatible change, right?  Can we do this backwards-compatibly?

bzbarsky-apple commented on src/app/clusters/diagnostic-logs-server/BDXDiagnosticLogsProvider.h:
```diff
@@ -77,7 +77,7 @@ class BDXDiagnosticLogsProvider : public bdx::Initiator
      * exchange and ends log collection.
      */
-    void Reset();
+    void Reset(CHIP_ERROR error);
```
Can we do this backwards-compatibly (e.g. by having error be optional, defaulting to something)?

Also, something needs to document the meaning of the "error" argument.

bzbarsky-apple commented on src/app/clusters/diagnostic-logs-server/DiagnosticLogsProviderDelegate.h:
```diff
@@ -59,9 +59,10 @@ class DiagnosticLogsProviderDelegate
      * returned from StartLogCollection.
      *
      * @param[in] sessionHandle  The unique handle for this log session returned from a call to StartLogCollection.
+     * @param[in] error A CHIP_ERROR value that indicates the reason for ending the log collection.
```
Should be made clear that this is allowed to be CHIP_NO_ERROR.

pimpalemahesh committed:
diagnostic-logs-cluster: make changes backword compatible

shubhamdp commented on src/app/clusters/diagnostic-logs-server/BDXDiagnosticLogsProvider.h:
```diff
@@ -76,8 +76,10 @@ class BDXDiagnosticLogsProvider : public bdx::Initiator
+     *                   It is permissible to pass CHIP_NO_ERROR to indicate normal termination.
      */
-    void Reset();
+    void Reset(CHIP_ERROR error = CHIP_NO_ERROR);
```
As this is a private API, no need to have a default value?

pimpalemahesh committed:
diagnostic-logs-cluster: remove default parameter from Reset() method

andy31415 reviewed (approved):
@pimpalemahesh even though it is explained in the linked issue, please update the summary of the issue to explain *WHY* these changes are being made and how it makes life better.

The reason for asking this is that these messages make their way in git log, and I would much rather understand the changes from the log than needing to also open an issue to follow up.

bzbarsky-apple commented on src/app/clusters/diagnostic-logs-server/DiagnosticLogsProviderDelegate.h:
```diff
@@ -59,9 +59,11 @@ class DiagnosticLogsProviderDelegate
      *
      */
-    virtual CHIP_ERROR EndLogCollection(LogSessionHandle sessionHandle) = 0;
+    virtual CHIP_ERROR EndLogCollection(LogSessionHandle sessionHandle, CHIP_ERROR error = CHIP_NO_ERROR) = 0;
```
This is still a breaking change.  Like I asked: do we want to cause a break here, or do we want to do this in a backwards-compatible manner?  That question was not answered.

pimpalemahesh replied:
@bzbarsky-apple I have overloaded the method in a backward-compatible manner. Please review.

bzbarsky-apple commented:
> The reason for asking this is that these messages make their way in git log

Note: the PR description does not end up in the git log.  The commit messages from the actual commits do....  Those should be describing why the change is being made.  And they are not.

pimpalemahesh committed:
diagnostic-logs-delegate: Add default implementation for EndLogCollection overload

	- Provides a default implementation for the EndLogCollection method with an additional error parameter.
	- This ensures backward compatibility, reduces the need for overriding in derived classes, and supports scenarios requiring context for log collection termination.

bzbarsky-apple commented on src/app/clusters/diagnostic-logs-server/DiagnosticLogsProviderDelegate.h:
```diff
@@ -63,6 +63,27 @@ class DiagnosticLogsProviderDelegate
+     *       sessionHandle.
+     *
+     */
+    virtual CHIP_ERROR EndLogCollection(LogSessionHandle sessionHandle, CHIP_ERROR error)
```
OK, but doing this forces someone who wants the new API to implement both, as you discovered.  I'd suggest something more like this:

```
virtual CHIP_ERROR EndLogCollection(LogSessionHandle sessionHandle) {
    return CHIP_NO_ERROR; /* Or return some error? */
}
virtual CHIP_ERROR EndLogCollection(LogSessionHandle sessionHandle, CHIP_ERROR error) 
{ 
    return EndLogCollection(sessionHandle);
}
```

and document that the two-arg form is what will be invoked by the actual diagnostic logs code.  Then old implementations act as they always have, new ones only need to implement the two-arg form.  A new implementation that does not override either form won't work right and won't get caught at compile time, but I think that's preferable to forcing API consumers to implement things twice.

pimpalemahesh replied:
@bzbarsky-apple You're right. Making EndLogCollection() default removes the need to redundantly override this method. I have updated the changes as per your suggestion. Thanks for the input. Please review.

pimpalemahesh committed:
diagnostic-logs-delegate: Add default implementation for EndLogCollection

- The one-argument EndLogCollection method is retained for backward compatibility.
- The new two-argument overloaded EndLogCollection method will serve as the primary
  method to be implemented in delegates.

bzbarsky-apple commented on examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp:
```diff
@@ -65,7 +65,7 @@ CHIP_ERROR LogProvider::GetLogForIntent(IntentEnum intent, MutableByteSpan & out
     VerifyOrReturnError(CHIP_NO_ERROR == err, err, outBuffer.reduce_size(0));
 
-    err = EndLogCollection(sessionHandle);
+    err = EndLogCollection(sessionHandle, err);
```
So just to be clear: this always passes CHIP_NO_ERROR for the second arg, right?

pimpalemahesh replied:
Yes. this is redundant I will remove it.

bzbarsky-apple replied:
Well, shouldn't we EndLogCollection even of CollectLog errored?

pimpalemahesh replied:
I realized that the err parameter is indeed necessary in this context. However, CHIP_NO_ERROR was not the appropriate value, so I have updated it to pass the correct error instead. Thank you for pointing this out.

bzbarsky-apple replied:
@pimpalemahesh If this line is reached, `err` is CHIP_NO_ERROR.  If it were not, we would have returned on line 66.

pimpalemahesh replied:
@bzbarsky-apple Apologies for the misunderstanding and for initially missing the changes. I've reviewed the issue and raised a PR with the necessary updates: [Pull Request](https://github.com/project-chip/connectedhomeip/pull/36952) Kindly review it at your convenience.

bzbarsky-apple replied:
@pimpalemahesh  Great, thanks!  Please see my comment on that PR, though.

bzbarsky-apple commented on src/app/clusters/diagnostic-logs-server/BDXDiagnosticLogsProvider.cpp:
```diff
@@ -191,7 +191,11 @@ void BDXDiagnosticLogsProvider::OnAckReceived()
+        err = mDelegate->EndLogCollection(mLogSessionHandle, err);
+        if (err == CHIP_ERROR_NOT_IMPLEMENTED)
+        {
+            ChipLogError(DeviceLayer, "EndLogCollection not implemented in the delegate. Falling back to default behavior.");
```
Which default behavior?  We do exactly the same thing no matter what EndLogCollection returns, raising questions about why it returns anything at all.....

pimpalemahesh replied:
@bzbarsky-apple I added this to notify the application side if EndLogCollection is not implemented, in case the user misses it, as we have provided default implementations for earlier virtual methods.
I will remove it.

bzbarsky-apple commented on src/app/clusters/diagnostic-logs-server/BDXDiagnosticLogsProvider.cpp:
```diff
@@ -279,7 +283,11 @@ void BDXDiagnosticLogsProvider::Reset()
+        CHIP_ERROR err = mDelegate->EndLogCollection(mLogSessionHandle, error);
+        if (err == CHIP_ERROR_NOT_IMPLEMENTED)
+        {
+            ChipLogError(DeviceLayer, "EndLogCollection not implemented in the delegate. Falling back to default behavior.");
```
Again, this does not seem to describe what we really do, which is ignore the return value entirely.

pimpalemahesh replied:
@bzbarsky-apple I have removed these unnecessary lines. Please review.

pimpalemahesh committed:
fix: remove unnecessary chip log statements

pimpalemahesh committed:
BDX: modify EndLogCollection parameter for internal log end requests
